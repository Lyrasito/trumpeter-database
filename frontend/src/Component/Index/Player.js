import "./Player.css";
import React from "react";
import ExpandedPlayer from "./ExpandedPlayer";
import { ReactComponent as CaretDown } from "../../svg/Caret-Down-2.svg";
import { ReactComponent as CaretUp } from "../../svg/Caret-Up.svg";

class Player extends React.Component {
  state = {
    playerShown: false,
  };

  handleClick = () => {
    this.setState({
      playerShown: !this.state.playerShown,
    });
  };

  render() {
    return (
      <div className="container ">
        <div className="player-line-container">
          <div
            key={this.props.player.id}
            className="player-line"
            onClick={this.handleClick}
          >
            <h5>{this.props.player.name}</h5>
            {this.state.playerShown ? (
              <CaretUp id="caret" />
            ) : (
              <CaretDown id="caret" />
            )}
          </div>
          <div id="shownPlayer">
            {this.state.playerShown && (
              <ExpandedPlayer player={this.props.player} />
            )}
          </div>
        </div>
      </div>
    );
  }
}

export default Player;
