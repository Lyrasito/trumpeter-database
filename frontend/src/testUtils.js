const fakePlayer = () => ({
  id: 1,
  name: "Fake Player",
  city: "Fake City",
  startYear: 1234,
  endYear: 5678,
});

const fakePlayers = () => [
  {
    id: 1,
    name: "Fake Player",
    city: "Fake City",
    startYear: 1234,
    endYear: 5678,
  },
  {
    id: 2,
    name: "Fake Player 2",
    city: "Fake City 2",
    startYear: 2345,
    endYear: 6789,
  },
];

const fakeAlbums = () => [
  {
    id: 1,
    title: "Fake Album 1",
    year: 1234,
    genre: "fakeGenre1",
    player_id: 1,
  },
  {
    id: 2,
    title: "Fake Album 2",
    year: 1234,
    genre: "fakeGenre2",
    player_id: 1,
  },
];

const fakeAlbum = () => ({
  id: 1,
  title: "Fake Album 1",
  year: 1234,
  genre: "fakeGenre1",
  player_id: 1,
});

const fakeGenres = () => ["fakeGenre1", "fakeGenre2"];

export { fakeAlbum, fakeAlbums, fakePlayer, fakeGenres, fakePlayers };
