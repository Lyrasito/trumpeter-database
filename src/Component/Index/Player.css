#hiddenAlbums {
  display: none;
}

#shownAlbums {
  display: flex;
  flex-wrap: wrap;
  width: 1280px;
  background-color: rgba(255, 255, 255, 0.1);
}

.albumsHeader {
  display: flex;
  height: 60px;
}

.clickButton {
  cursor: pointer;
  text-decoration: underline;
  margin-left: 40px;
}
#album-header {
  margin-left: 20px;
}

.img {
  width: 320px;
  height: 320px;
  object-fit: cover;
}
img {
  width: 300px;
  height: auto;
}
.album-image {
  width: 240px;
  height: 240px;
}
.player-container {
  display: flex;
  justify-content: space-between;
  padding-bottom: 40px;
}

.player-album-container {
  border-bottom: 4px solid #ffeeb4;
}
.career-span {
  margin-right: 80px;
  margin-left: 30px;
  width: 400px;
  padding-top: 0px;
  vertical-align: super;
  margin-top: -40px;
}
.city-genres {
  margin-top: -40px;
}
.small-title {
  color: #ffeeb4;
}

.link {
  color: white;
}

.album {
  width: 240px;
  margin: 38px;
}

.album-title {
  height: 35px;
  width: 240px;
  color: #ffffff;
  font-family: "Work Sans";
  font-size: 24px;
  font-weight: bold;
  letter-spacing: 0;
  line-height: 35px;
}

#genreFilter {
  display: flex;
  margin-left: 780px;
  height: 40px;
  margin-top: 10px;
  justify-content: space-between;
}
.genreSelect {
  box-sizing: border-box;
  height: 40px;
  width: 100px;
  border: 3px solid #666666;
  border-radius: 30px;
  background-color: #ffffff;
  display: block;
}
.filterButton {
  height: 40px;
  width: 100px;
  border-radius: 30px;
  background-color: #ffc600;
  cursor: pointer;
  color: #000000;
  font-family: "WorkSans";
  font-size: 20px;
  letter-spacing: 0;
  line-height: 35px;
  margin-left: 10px;
}
