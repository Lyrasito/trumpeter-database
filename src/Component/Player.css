#hidden {
  display: none;
}

#shown {
  display: inline-block;
}

.clickButton {
  cursor: pointer;
}
