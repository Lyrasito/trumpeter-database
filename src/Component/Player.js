import React from "react";

class Player extends React.Component {
  render() {
    return (
      <div>
        <h1>Name: {this.props.name}</h1>
        <h2>City: {this.props.city}</h2>
        <h2>
          Career Span: {this.props.startYear}-{this.props.endYear}
        </h2>
      </div>
    );
  }
}

export default Player;
