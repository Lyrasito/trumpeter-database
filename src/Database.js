const fetch = require("node-fetch");

const baseUrl = "http://localhost:4000/api";

const Database = {
  async getPlayers() {
    const url = `${baseUrl}/players`;
    const response = await fetch(url);
    const jsonResponse = await response.json();
    const players = await jsonResponse.players.map((player) => ({
      id: player.id,
      name: player.name,
      city: player.city,
      startYear: player.start_year,
      endYear: player.end_year,
      image: player.image,
    }));
    return players;
  },

  async searchPlayers(city, year, genre) {
    let url = "";
    if (city && !year && !genre) {
      url = `${baseUrl}/players/search?city=${city}`;
    } else if (year && !city && !genre) {
      url = `${baseUrl}/players/search?year=${year}`;
    } else if (genre && !city && !year) {
      url = `${baseUrl}/players/search?genre=${genre}`;
    } else if (city && year && !genre) {
      url = `${baseUrl}/players/search?city=${city}&year=${year}`;
    } else if (city && genre && !year) {
      url = `${baseUrl}/players/search?city=${city}&genre=${genre}`;
    } else if (genre && year && !city) {
      url = `${baseUrl}/players/search?genre=${genre}&year=${year}`;
    } else if (city && year && genre) {
      url = `${baseUrl}/players/search?city=${city}&year=${year}&genre=${genre}`;
    }
    const response = await fetch(url);
    const jsonResponse = await response.json();
    const players = await jsonResponse.players.map((player) => ({
      id: player.id,
      name: player.name,
      city: player.city,
      startYear: player.start_year,
      endYear: player.end_year,
      image: player.image,
    }));
    return players;
  },

  async searchByName(name) {
    const url = `${baseUrl}/players/search?name=${name}`;
    const response = await fetch(url);
    const jsonResponse = await response.json();
    const players = await jsonResponse.players.map((player) => ({
      id: player.id,
      name: player.name,
      city: player.city,
      startYear: player.start_year,
      endYear: player.end_year,
      image: player.image,
    }));
    return players;
  },

  async getPlayerAlbums(player) {
    const url = `${baseUrl}/players/${player.id}/albums`;
    const response = await fetch(url);
    const jsonResponse = await response.json();
    const albums = await jsonResponse.albums.map((album) => ({
      id: album.id,
      title: album.title,
      year: album.year,
    }));
    return albums;
  },

  async getPlayerGenres(player) {
    const url = `${baseUrl}/players/${player.id}/albums/genres`;
    const response = await fetch(url);
    const jsonResponse = await response.json();
    return jsonResponse.genre;
  },

  async addPlayer(name, city, startYear, endYear) {
    const url = `${baseUrl}/players`;
    const response = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        player: {
          name: name,
          city: city,
          startYear: startYear,
          endYear: endYear,
        },
      }),
    });

    const jsonResponse = await response.json();
    return jsonResponse.player;
  },

  async addAlbum(playerId, title, year, genre) {
    const url = `${baseUrl}/players/${playerId}/albums`;
    const response = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        album: {
          title: title,
          year: year,
          genre: genre,
          playerId: playerId,
        },
      }),
    });
    const jsonResponse = await response.json();
    console.log(jsonResponse);
  },
};

module.exports = Database;

/*  */
