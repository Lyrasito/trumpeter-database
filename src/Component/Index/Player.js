import React from "react";
import "./Player.css";
import Spotify from "../../Spotify";
import PlayerPlaceholder from "../../svg/PlayerPlaceholder.svg";

class Player extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      albumsHidden: true,
      albumLink: [],
      genre: "",
    };
    this.showAlbums = this.showAlbums.bind(this);
    this.getSpotify = this.getSpotify.bind(this);
  }
  async showAlbums() {
    if (this.state.albumsHidden) {
      this.setState({
        albumsHidden: false,
      });
    } else {
      this.setState({
        albumsHidden: true,
      });
    }
    for (let i = 0; i < this.props.player.albums.length; i++) {
      let link = await this.getSpotify(
        `${this.props.player.albums[i].title} ${this.props.player.name}`
      );
      let links = this.state.albumLink;
      links.push(link);
      this.setState({
        albumLink: links,
      });
    }
  }
  async getSpotify(term) {
    const link = await Spotify.searchAlbum(term);
    return link;
  }

  render() {
    if (this.props.player) {
      return (
        <div className="player-album-container">
          <h1>{this.props.player.name}</h1>
          <div className="player-container">
            <img
              src={
                this.props.player.image
                  ? this.props.player.image
                  : PlayerPlaceholder
              }
              alt=""
              className="img"
            />
            <div className="career-span">
              <h5 className="small-title">Career Span:</h5>
              <h3>
                {this.props.player.startYear}-{this.props.player.endYear}
              </h3>
            </div>
            <div className="city-genres">
              <h5 className="small-title">City: </h5>
              <h3>{this.props.player.city}</h3>

              <h5 className="small-title">Genres:</h5>
              <h3>{this.props.player.genres.join(", ")}</h3>
            </div>
          </div>
          <div className="albumsHeader">
            <div className="showHide">
              <p className="small-title" id="album-header">
                Albums
              </p>
              <p onClick={this.showAlbums} className="clickButton">
                Show/Hide
              </p>
            </div>
            <div id={this.state.albumsHidden ? "hiddenAlbums" : "genreFilter"}>
              <label>Filter by genre:</label>
              <select className="genreSelect" onChange={this.props.getGenre}>
                <option>Select</option>
                {this.props.player.genres.map((genre, index) => {
                  return (
                    <option value={genre} key={index}>
                      {genre}
                    </option>
                  );
                })}
              </select>
              <button
                type="submit"
                className="filterButton"
                onClick={this.props.filterByGenre}
              >
                Filter
              </button>
            </div>
          </div>
          <div id={this.state.albumsHidden ? "hiddenAlbums" : "shownAlbums"}>
            {this.props.albums.map((album, index) => {
              return (
                <div key={album.id} className="album">
                  <img
                    src="./img/album-default.png"
                    className="album-image"
                    alt=""
                  />
                  <h5 className="album-title">{album.title} </h5>{" "}
                  <label>{album.year}</label>
                  <br />
                  <label>Genre: {album.genre}</label>
                  <a
                    href={this.state.albumLink[index]}
                    className="link"
                    target="_blank"
                  >
                    <p>Open with Spotify</p>
                  </a>
                </div>
              );
            })}
          </div>
        </div>
      );
    } else {
      return null;
    }
  }
}

export default Player;
