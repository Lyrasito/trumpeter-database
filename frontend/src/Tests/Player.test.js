import Player from "../Component/Index/Player";
import React from "react";
import toJSON from "enzyme-to-json";
import { shallow } from "enzyme";
import wait from "waait";
import Database from "../Database";
import { fakePlayer, fakeGenres } from "../testUtils";

describe("<Player />", () => {
  it("renders properly", () => {
    const wrapper = shallow(<Player player={fakePlayer()} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
  it("renders an <ExpandedPlayer/> component when a player is clicked", async () => {
    const wrapper = shallow(<Player player={fakePlayer()} />);
    Database.getPlayerGenres = jest.fn().mockResolvedValue(fakeGenres());

    const playerLine = wrapper.find(".player-line");
    playerLine.simulate("click");
    await wait();
    wrapper.update();
    expect(wrapper.find("ExpandedPlayer")).toHaveLength(1);
  });
});
