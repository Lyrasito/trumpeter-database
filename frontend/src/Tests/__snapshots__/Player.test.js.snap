// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Player /> renders properly 1`] = `
<div
  className="container "
>
  <div
    className="player-line-container"
  >
    <div
      className="player-line"
      key="1"
      onClick={[Function]}
    >
      <h5>
        Fake Player
      </h5>
      <ForwardRef(SvgCaretDown2)
        id="caret"
      />
    </div>
    <div
      id="shownPlayer"
    />
  </div>
</div>
`;
